index.js

```diff
@@ -1,104 +1,110 @@
+			if (!user || !user.fullname) {
+				throw new Error();
+			}
 
-	const envVar = getEnvVar();
+			return user.fullname;
+		});
+}
 
-	if (envVar) {
-		fullname = envVar;
-		return Promise.resolve(fullname);
-	}
+function checkGit() {
+	return execa('git', ['config', '--global', 'user.name']).then(res => {
+		if (!res.stdout) {
+			throw new Error();
+		}
 
-	return Promise.resolve().then(() => {
-		fullname = require('rc')('npm')['init-author-name'];
+		return res.stdout;
+	});
+}
 
+function checkOsaScript() {
+	return execa.stdout('osascript', ['-e', 'long user name of (system info)'])
+		.then(fullname => {
+			if (!fullname) {
+				throw new Error();
+			}
+
+			return fullname;
+		});
+}
+
+function checkWmic() {
+	return execa.stdout('wmic', ['useraccount', 'where', 'name="%username%"', 'get', 'fullname'])
+		.then(stdout => {
+			const fullname = stdout.replace('FullName', '');
+
+			if (!fullname) {
+				throw new Error();
+			}
+
+			return fullname;
+		});
+}
+
+function checkGetEnt() {
+	return execa.shell('getent passwd $(whoami)').then(res => {
+		const fullname = (res.stdout.split(':')[4] || '').replace(/,.*/, '');
 		if (!fullname) {
-			return fallback().then(() => fullname);
+			throw new Error();
 		}
 
 		return fullname;
-	}).catch(fallback).then(() => fullname).catch(() => {});
-};
+	});
+}
 
 function fallback() {
 	if (process.platform === 'darwin') {
-		return passwdUser()
-			.then(user => {
-				fullname = user.fullname;
-				return fullname;
-			})
-			.catch(() => {
-				return execa.stdout('osascript', ['-e', 'long user name of (system info)']).then(stdout => {
-					fullname = stdout;
-					return fullname;
-				});
-			});
+		return pAny([checkPasswd(), checkOsaScript()]);
 	}
 
 	if (process.platform === 'win32') {
-		// try git first since fullname is usually not set by default in the system on Windows 7+
-		return execa('git', ['config', '--global', 'user.name'])
-			.then(res => {
-				fullname = res.stdout;
-
-				if (!fullname) {
-					throw new Error();
-				}
-			})
-			.catch(() => {
-				return execa.stdout('wmic', ['useraccount', 'where', 'name="%username%"', 'get', 'fullname'])
-					.then(stdout => {
-						fullname = stdout.replace('FullName', '');
-						return fullname;
-					});
-			});
+		return pAny([checkGit(), checkWmic()]);
 	}
 
-	return passwdUser()
-		.then(user => {
-			fullname = user.fullname;
+	return pAny([checkPasswd(), checkGetEnt(), checkGit()]);
+}
 
-			if (!fullname) {
-				throw new Error();
-			}
-		})
-		.catch(() => {
-			return execa.shell('getent passwd $(whoami)')
-				.then(res => {
-					fullname = (res.stdout.split(':')[4] || '').replace(/,.*/, '');
-
-					if (!fullname) {
-						throw new Error();
-					}
-				});
-		})
-		.catch(() => {
-			return execa.stdout('git', ['config', '--global', 'user.name'])
-				.then(stdout => {
-					fullname = stdout;
-					return fullname;
-				});
-		});
+function getFullName() {
+	return pAny([checkEnv(), checkAuthorName()])
```

Comment:
No point in using pAny here. `checkEnv` will always finish first. Just chain it.